rustworkx-core/src/coloring.rs

```diff
@@ -141,21 +176,190 @@ where
     Ok(colors)
 }
 
+/// Data associated to nodes for the greedy coloring algorithm
+/// using the "saturation first" strategy: always picking the node that
+/// has the largest number of different colors already assigned to its
+/// neighbors, and, in case of a tie, the node that has the largest number
+/// of uncolored neighbors.
+#[derive(Clone, Eq, PartialEq)]
+struct SaturationStrategyData {
+    // degree of a node: number of neighbors without color
+    degree: usize,
+    // saturation degree of a node: number of colors used by neighbors
+    saturation: usize,
+}
+
+impl Ord for SaturationStrategyData {
+    fn cmp(&self, other: &SaturationStrategyData) -> Ordering {
+        self.saturation
+            .cmp(&other.saturation)
+            .then_with(|| self.degree.cmp(&other.degree))
+    }
+}
+
+impl PartialOrd for SaturationStrategyData {
+    fn partial_cmp(&self, other: &SaturationStrategyData) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+fn inner_greedy_node_color_strategy_saturation<G, F, E>(
+    graph: G,
+    mut preset_color_fn: F,
+) -> Result<DictMap<G::NodeId, usize>, E>
+where
+    G: NodeCount + IntoNodeIdentifiers + IntoEdges + NodeIndexable,
+    G::NodeId: Hash + Eq + Send + Sync,
+    F: FnMut(G::NodeId) -> Result<Option<usize>, E>,
+{
+    let mut colors: DictMap<G::NodeId, usize> = DictMap::with_capacity(graph.node_count());
+
+    let mut nbd_colors: HashMap<G::NodeId, HashSet<usize>> =
+        HashMap::with_capacity(graph.node_count());
+    for k in graph.node_identifiers() {
+        nbd_colors.insert(k, HashSet::new());
+    }
+
+    let mut pq: PriorityQueue<G::NodeId, SaturationStrategyData> =
+        PriorityQueue::with_capacity(graph.node_count());
+
+    // Handle preset nodes
+    for k in graph.node_identifiers() {
+        if let Some(color) = preset_color_fn(k)? {
+            colors.insert(k, color);
+            for v in graph.neighbors(k) {
+                nbd_colors.get_mut(&v).unwrap().insert(color);
+            }
+        }
+    }
+
+    // Add non-preset nodes to priority queue
+    for k in graph.node_identifiers() {
+        if colors.get(&k).is_none() {
+            let degree = graph
+                .neighbors(k)
+                .filter(|v| colors.get(v).is_none())
+                .count();
+            let saturation = nbd_colors.get(&k).unwrap().len();
+            pq.push(k, SaturationStrategyData { degree, saturation });
+        }
+    }
+
+    // Greedily process nodes
+    while let Some((k, _)) = pq.pop() {
+        let neighbor_colors = nbd_colors.get(&k).unwrap();
+        let mut current_color: usize = 0;
+        loop {
+            if !neighbor_colors.contains(&current_color) {
+                break;
+            }
```

Comment:
Could you do this instead?

```suggestion
        while !neighbor_colors.contains(&current_color) {
```